pgxn/neon/pagestore_smgr.c

```diff
@@ -1152,18 +1152,34 @@ neon_read_at_lsn(RelFileNode rnode, ForkNumber forkNum, BlockNumber blkno,
 		if (n_prefetch_requests > 0)
 		{
 			/* Combine all prefetch requests with primary request */
-			page_server->send((NeonRequest *) & request);
+			if (request_latest)
+			{
+				/*
+				 * Use maximal LSN of all prefetch requests to minimize probability of rejection of prefetch
+				 * responses. Prefetch response is rejected if Max(prefetch_lsn, PageGetLSN(page)) < request_lsn.
+				 * Assume that there are two pages P1 and P2 with LSNs 100 and 200 and them are swapped out.
+				 * Then we want to load P1 and prefetch P2. If we use LSN 100 for all prefetched pages,
+				 * then pageserver will not wait until it receives WAL till LSN=200 and returns old image of P2.
+				 * So prefetch response for P2 will be rejected.
```

Comment:
Thanks for this example, that's very helpful. I still don't fully understand this though. Why would we use the same LSN for P1 and P2? That's not what we did before this PR either. Without this PR, we would use LSN 100 in the request for P1 and LSN 200 in the request for P2, and set prefetch_lsn=200.

I don't really understand the `if (Max(prefetch_lsn, PageGetLSN(page)) >= request_lsn)` above, where we check if a prefetched response is still valid. If we got LSN 1000 from last-written LSN cache,  but received a page with LSN 2000, is that really OK? Seems like our last-written LSN cache was wrong, and we should've used a higher request LSN in the first place. Print a warning at least?

I think we should remember the LSN used in each prefetch request separately. https://github.com/neondatabase/neon/pull/2687 has a per-request field for that, I think that's the most straightforward approach (although now that I look at that more closely, I don't think https://github.com/neondatabase/neon/pull/2687 sets the LSN field correctly; thank you for making me have a closer look at that!).

Let's walk through another similar example: Let's assume that the last-written LSN cache is full but doesn't contain any pages of interest, and the global last-written LSN is 100. We issue prefetch requests for P1 and P2 with LSN 100, and set prefetch_lsn to 100. Next, we modify some unrelated page, and that kicks another page out of the last-written LSN cache, which bumps the global last-written LSN to 101. Then, we perform the actual read for P1, with LSN 101. We find the response for P1 in the prefetch buffers, but prefetch_lsn is 100, so it is rejected.

That seems like a problem. If it's a large sequential scan, and you also keep modifying stuff, we might reject every prefetched page. Perhaps we should insert the entry to the last-written LSN cache when we issue the prefetch request, so that the real request will use the same LSN (unless the page is actually modified in between).

Reply:
> Why would we use the same LSN for P1 and P2?

It is because of this `(Max(prefetch_lsn, PageGetLSN(page)) >= request_lsn)` check.and the way how `prefetch_lsn` is calculated (as max of last written LSNs of all prefetched pages). If we request P1 with LSN=100 but then set `prefetch_lsn-200` then in this check we will compare `request_lsn` with 200. But it is not correct.
Please notice that only original page is locked and protected by concurrent access and modification (through pinned shared buffer). All prefetche pages can be concurrently modified by other backends. So if we load page with `latest` flag, we can assume that we really get the latest version of the page. But it is not true for prefetched pages. If we request page with `request_lsn=100`, then we can get for example page with LSN=99 while there is already new version of this page with LSN=150. If then in the up-to-date check of prefetch result we compare `max(200,99) >= 150` and accept deterirated response. This is why I request all pages with lsn=max(last evicted lsns)

Reply:
> If we got LSN 1000 from last-written LSN cache, but received a page with LSN 2000, is that really OK?

It can not happen for original page (protected by shared buffer lock), but can happen for preftched pages.

Reply:
> I think we should remember the LSN used in each prefetch request separately. 

It is also correct. I used the same approach in the branch `local_prefetch_continuous` (when prefetch requests are sent separately). But it just increases probability of prefetch response rejection. And requesting all prefetch pages with max(last_writte_lsn) reduces this probability without almost no negative impact. Tt may cause pageserver to wait for lsn a little bit earlier than it is actually needed, but prevents sending deteriorated result  and breaking prefetch pipe which seems to be much ore critical.

Reply:
> Then, we perform the actual read for P1, with LSN 101. We find the response for P1 in the prefetch buffers, but prefetch_lsn is 100, so it is rejected.

Yes, exactly!  But unfortunately there is no way prevent it. We do not know whether this modified page is actually unrelated or not. If page is absent in last written lsn cache then we cn make only conservative assumption that it was modified at LSN=101. Fortunately probability of such false prefetch rejection is relatively small because of LRU. And the role of `prefetch_lsn` is to reduce this probability. We can certainly issue prefetch with `request_lsn=current_wal_position`. It will reduce probability of prefetch rejection even more.  But it will cause pageserver to wait until all WAL is received which most likely is not needed for the reconstruction fo the requested pages.

Please notice that it is not only my "speculations": actually this change `request_lsn=max(last written lsns)` gives the most noticeable improvement of seqscan of pgbench_accounts after normal pgbench iteration (performing random updates).

Reply:
> Perhaps we should insert the entry to the last-written LSN cache when we issue the prefetch request, so that the real request will use the same LSN (unless the page is actually modified in between).

Yes, it may be also useful to reduce probability of false prefetch response rejection. But it may also flush last written lsn cache, because now we will store in it not only LSNs of most recently updated pages.

Reply:
I think more about your suggestion to store last written lsn for each prefetch pages and it seems to be not so useful.
First of all we do not know whether GetLastWrittenLsn returns precise LSN of this page located in the cache or global last written lsn (max of LSNs of all evicted pages). There is no sense in storing global LSN. But storing precise last written LSN also seems to be useful. If this page was present in the cache then access will move this entry to the head of LRU list and probability that it will be thrown away before prefetched page is accessed is very small.

Reply:
Created https://github.com/neondatabase/neon/issues/2857 to track this issue